### PR TITLE
Allow keras-preprocessing to process bigger datasets of text documents

### PR DESCRIPTION
**Pull request by cdemutiis**
This pull request aims at giving the user the ability to pick the NumPy data type of the output NumPy matrix when using the Tokenizer's method called texts_to_matrix in keras_preprocessing/text.py. This can result in a much more efficient memory usage and in the possibility of processing bigger datasets of text documents. 

Example:
In the case of a simple binary BOW One-Hot encoding, the output NumPy matrix does not need to have the np.float64 data type as all of its values are either 0 or 1. Using the np.int8 type would lead to potentially huge savings in memory, i.e. 32 - 25 = 7 bytes per matrix element. 

P.S. 
I introduced a conditional block that reverts the NumPy data type to np.float64 whenever the NumPy data type input by the user leads to loss of information in the current mode in sequences_to_matrix(). If that is the case, an appropriate warning gets displayed.

### CODE REVIEW

**Comment by cdemutiis** [comment]
The line of code that usually results in a Python Memory Error for very large matrices is 
`x = np.zeros((len(sequences), num_words))` in the Tokenizer's method called texts_to_matrix in keras_preprocessing/text.py

File: tests/text_test.py [hunk 1]
```diff
@@ -248,6 +248,38 @@ def test_sequences_to_texts():
...
 
+def test_sequences_to_matrix():
+    # Tests for the 'count' mode NumPy data type
+    int_types = [
+                (8, np.uint8),
+                (16, np.uint16),
+                (32, np.uint32),
+                (64, np.uint64)
+                ]
+    for i in range(len(int_types) - 2):
```
Comment (cdemutiis) [inline tests/text_test.py]: @Dref360 the reason why this for loop only goes up to -2 is due to the fact that generating strings with a number of words in the order of 2**32 or more takes too much memory.

**Comment by cdemutiis** [comment]
As a side note, it would be great to remove the **following warnings** I noticed when running the tests
  
`tests/image_test.py::TestImage::()::test_image_data_generator_invalid_data
  /Users/claudiod/Desktop/Git_Projects_claudio_dm/keras_fork/keras-preprocessing/keras_preprocessing/image/image_data_generator.py:340: UserWarning: This ImageDataGenerator specifies `zca_whitening` which overrides setting of`featurewise_std_normalization`.
    warnings.warn('This ImageDataGenerator specifies '

tests/image_test.py::TestImage::()::test_image_data_generator
  /Users/claudiod/Desktop/Git_Projects_claudio_dm/keras_fork/keras-preprocessing/keras_preprocessing/image/image_data_generator.py:340: UserWarning: This ImageDataGenerator specifies `zca_whitening` which overrides setting of`featurewise_std_normalization`.
    warnings.warn('This ImageDataGenerator specifies '

tests/image_test.py::TestImage::()::test_image_data_generator_fit
  /Users/claudiod/Desktop/Git_Projects_claudio_dm/keras_fork/keras-preprocessing/keras_preprocessing/image/image_data_generator.py:340: UserWarning: This ImageDataGenerator specifies `zca_whitening` which overrides setting of`featurewise_std_normalization`.
    warnings.warn('This ImageDataGenerator specifies '

tests/image_test.py::TestImage::()::test_batch_standardize
  /Users/claudiod/Desktop/Git_Projects_claudio_dm/keras_fork/keras-preprocessing/keras_preprocessing/image/image_data_generator.py:340: UserWarning: This ImageDataGenerator specifies `zca_whitening` which overrides setting of`featurewise_std_normalization`.
    warnings.warn('This ImageDataGenerator specifies '

-- Docs: http://doc.pytest.org/en/latest/warnings.html`

**Comment by Dref360** [comment]
From a UX standpoint, would the user expect to get a different type depending on something as variable as the count?

**Comment by Dref360** [comment]
I do think that adding a `dtype` args ,which by default would be int_, would be more user-friendly.

**Comment by cdemutiis** [comment]
> I do think that adding a `dtype` args ,which by default would be int_, would be more user-friendly.

Thank you for your prompt reply @Dref360! I see the merits of your point in terms of giving the user the choice rather than making it behind the curtains. However, wouldn't it be better if the optimal data type is picked depending on the 'mode' used in order to guarantee memory savings and decrease the risk of the method freezing (as it happened to me in the past)? Could you point to a practical disadvantage of this approach? As a side note, I believe that operations among different NumPy data types are in general allowed and should inherit the largest data type. 

P.S. 
- In the current version of the pull request, float64 is not necessarily the optimal data type for "tfidf" and "freq". I essentially took care of the 'binary' and 'count' modes.

**Comment by fchollet** [comment]
Closing outdated PR. Note that the Keras Preprocessing repository is no longer in use, instead please just use the dataset utilities provided in `tf.keras.utils` ([see docs](https://keras.io/api/data_loading/)).
